### ISSUE TITLE
New item

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8368]
As a DAWSON user, so that I have confidence that the data I'm working with is correct, I need DAWSON to maintain data integrity in the event of a race condition occurring.

Since launch, we have seen race conditions between operations in DAWSON present a variety of symptoms. These include documents incorrectly being recorded as served or unserved, transactions not entirely completing out their code path, and partially committed records in the database. Having data represented incorrectly in the system can have serious ramifications not just for internal Court processes but for the USTC case and its parties, who rely on DAWSON as the source of truth for this information.

Recent investigation has hypothesized that these problems occur when a race condition is present in the system and multiple operations or processes are attempting to act on shared or identical data. In order to better maintain the integrity of our data and have confidence that it accurately represents the truth, we need to fortify DAWSON against these conditions. Below in the Notes section are the results of recent analysis done on several instances of this in production DAWSON to help identify potential solutions for this problem.

## Acceptance Criteria
Spike Objectives:
* Document list of possible solution approaches
* Document workflow/code paths that call UpdateCaseAndAssociations
* Write a test to simulate the race condition, and red -> green it. (that can be the experiment)
  - Two different operations start with the same copy of the case.
  - One finishes first and calls updateCaseAndAssociations
  - The other finishes second and calls updateCaseAndAssociations
  - The resultant case has both changes saved

## Notes
* Answer to question "Who wins when multiple operations try to update the case?":
  - Technically speaking, they should *all* win. We just want to ensure that each change is made in an orderly fashion, with integrity (i.e., the change is committed in full, or prohibited from being made if the case data would otherwise prevent doing so)
* Timebox: 2 weeks?
* Please include @mmarcotte and/or @jimlerza and UX in mobbing 

From @mmarcotte - 

https://ustaxcourt.slack.com/archives/CD4PNKEPK/p1669826386714809

Suggestions made at refinement:

* Working group dedicated to analysis/scope
* Spike to understand breadth of problem, brainstorm approaches
* Narrow initial scope to the reproducible case noted herein to enable quick experimentation - UpdateCaseAndAssociations

## Story Definition of Ready (Created on 9/26/21)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ]  Be immediately actionable for the development team
- [ ]  Be focused on users
- [ ]  Include a narrative. “As a ____, so that I can ______, I need _____.”
- [ ]  Include user-centric acceptance criteria that expresses the users' needs
- [ ]  Include user-centric testing scenarios
- [ ]  Include preconditions (story dependencies)
- [ ]  Not include design or technical solutions
- [ ]  Include non-standard testing, deployment, documentation, browser, mobile or security requirements, as well as any requirements for 3rd-party integrations

Process: Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 5-19-22)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [ ] Automated test scripts have been written, including visual tests for newly added PDFs
 - [ ] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [ ] Verify that language for docket record for internal users and external users is identical
 - [ ] New screens have been added to pa11y scripts
 - [ ] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [ ] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [ ] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [ ] Interactors should validate entities before calling persistence methods
 - [ ] Code refactored for clarity and to remove any known technical debt
 - [ ] Acceptance criteria for the story has been met
 - [ ] If there are special instructions in order to deploy into the next environment, add them as a comment in the story
 - [ ] Reviewed by UX on a deployed environment.
 - [ ] Reviewed by PO on a deployed environment. Can be deployed to the Court's `test` environment if prod-like data is required. Otherwise deployed to any `experimental` environment.
 - [ ] Deployed to the Court's `staging` environment.

**Comment by mmarcotte** [comment]
# UpdateCaseAndAssociations Race Condition

When two operations run concurrently that both make use of `updateCaseAndAssociations`, they are likely to produce unpredictable or undesirable results on the Case and/or its associated records. This is because both retrieve an original copy of the Case, perform some operations on that copy, and then save their updated copy one after another. The second operation to save may overwrite or undo changes the operation had saved because they are different than the one it is saving.

These could be requests coming in from Respondent, the Party, or the Court. Any two requests that call `updateCaseAndAssociations` on the same Docket Number and are running at the same time are potentially problematic.

## Reproduction steps

I've created a branch where `updateDocketEntryMetaInteractor` has a sleep of ten seconds after it fetches a copy of the Case from the Database. This allows us to start that request, get a copy of the case, do something else that updates the case (like File and Serve a Court Issued Document) and observe the race condition.

1. Checkout [check-race-condition](https://github.com/ustaxcourt/ef-cms/tree/check-race-condition)
2. Start up your API and Client
3. Ensure you have a case where the Petition is served (I served the petition as Petitions Clerk on 102-22)
4. Login as a Docket Clerk
5. Navigate to your case
6. Create a Draft Order, apply a signature.
7. Click Add Docket Entry, put in a Date, and stop there
8. In another tab, Edit the meta of the Notice of Receipt of Petition. You may have to specify a Document Type (I did Notice) and put in  something for "What is this notice for?"
9. Now click Save, switch over to the other tab, and click Save and Serve.
10. The Add Docket Entry operation should finish first, and when it does you'll see it on the docket record and out of the Drafts tab.
11. Then the Update Docket Entry Meta request should finish second, and when it does you'll see the Court Issued Document returned to Drafts -- even though it has a Served Stamp on the PDF.

**Comment by mmarcotte** [comment]
https://user-images.githubusercontent.com/5023502/206783914-54fbd75c-2ef0-42ac-b5bc-a1bd3d7fc7e3.mp4

**Comment by codyseibert** [comment]
Some thoughts on research today:

I was able to fix the particular race condition detailed above by removing the call to updateCaseAndAssociations and instead call the individual updateCase and updateDocketEntry persistence methods directly.  The reason this prevents the original issue is because update case and associations will fetch the current state of the case and compare it against what the lambda is about to update.  If the lambda has an old copy of the docket entry (for example, when it is still in draft), but the lambda also sees there is a new copy in dynamo (docket entry was served), the code thinks the interactor was trying to set the served document back to a draft state, even though our code path has nothing to do with the docket entry in draft.

I agree with Mike that the best solution is to stop using updateCaseAndAssocations.  This method is doing TOO much complex logic in my opinion.  This complexity is resulting in the race conditions we are seeing.  We might be able to find ways to refactor the updateCaseAndAssociations to make it smarter, but I think we have a "omniscient function" which is a code smell.

Another issue we have in our application is that we call applicationContext.getPersistenceGateway().updateCase() which will OVERWRITE the existing case with the data we have in the lambda.  Because we do a full overwrite on records, there is a higher chance we will overwrite unrelated fields that could have been updated by other interactors by other users around the same time.  I think the best solution would be to only update the individual properties that have changed instead of the entire object, but finding an implementation for this is easier said than done.

I started adding some `patch` methods to our dynamodb client server which will allow us to easily update specific fields on records instead of overwriting them.  I need to discuss with others a good solution on how to keep track of "dirty" fields on the entity so that we can easily just have persistence know what fields have changed after we start modifying the entities.

**Comment by codyseibert** [comment]
```
participant userA
participant fileAndServeCourtIssuedDocumentInteractor
participant dynamodb
participant updateDocketEntryMetaInteractor
participant userB
userA->fileAndServeCourtIssuedDocumentInteractor: makes request
userB->updateDocketEntryMetaInteractor: makes request
dynamodb->updateDocketEntryMetaInteractor: fetch case
note over updateDocketEntryMetaInteractor: some delay occurs
dynamodb->fileAndServeCourtIssuedDocumentInteractor: fetch case
note over fileAndServeCourtIssuedDocumentInteractor: adds DocketEntry O to CaseEntity
note over fileAndServeCourtIssuedDocumentInteractor: updateCaseAndAssociations
fileAndServeCourtIssuedDocumentInteractor->dynamodb: overwrite Case
fileAndServeCourtIssuedDocumentInteractor->dynamodb: overwrite DocketEntry O\n(removes draft state)
note over updateDocketEntryMetaInteractor: update DocketEntry NOT metadata
note over updateDocketEntryMetaInteractor: updateCaseAndAssociations
dynamodb->updateDocketEntryMetaInteractor: fetches full case
note over updateDocketEntryMetaInteractor: diff between old and new case
note over updateDocketEntryMetaInteractor: thinks DocketEntry O and NOT changed
updateDocketEntryMetaInteractor->dynamodb: overwrite case
updateDocketEntryMetaInteractor->dynamodb: update DocketEntry O
updateDocketEntryMetaInteractor->dynamodb: update DocketEntry NOT
```

![Screen Shot 2023-01-19 at 12.13.01 PM.png](https://images.zenhubusercontent.com/5bbf6faf4368060246bbe5ce/7b38b8be-d341-4c69-8462-bc91add07001)

**Comment by codyseibert** [comment]
```
participant userA
participant updateCaseContextInteractor
participant dynamodb
participant updateDocketEntryMetaInteractor
participant userB
userA->updateCaseContextInteractor: changes status and title
userB->updateDocketEntryMetaInteractor: makes request
dynamodb->updateCaseContextInteractor: fetch case
dynamodb->updateDocketEntryMetaInteractor: fetch case
note over updateCaseContextInteractor: sets status to ready for trial
updateCaseContextInteractor->dynamodb: saves case
note over updateDocketEntryMetaInteractor: add docket entry to pending report
updateDocketEntryMetaInteractor->dynamodb: overwrite case
note over dynamodb: case status and title reverted
```
![Screen Shot 2023-01-19 at 4.03.39 PM.png](https://images.zenhubusercontent.com/5bbf6faf4368060246bbe5ce/5d9b4f98-5046-4f44-9d2a-f3a91834b075)

**Comment by codyseibert** [comment]
By doing patch writes, I think we will reduce the chance of race conditions.  If the goal is 100% race condition free updates, we will need to apply a lastUpdated or incrementing version number to our entities so that we can add in dynamo conditional updates to have our api fail for one of the requests.  Engineering a solution for this would require figuring out what a user would see if they were to make a request and their request failed due to a race condition.  Do we roll back any changes we made on the work items, cases, messages, docket entries etc?  What happens for endpoints that require PDF generation or emails to be sent?

We'd also need to figure out what we'd do if a race condition occur when doing a system related change, such as the cron job which sets cases ready for trial, or any other batch updating type of logic, such as when a user changes their email which will loop through every single case.  We can't just have those processes fail, so we may need to write a conflict resolution strategy for those instances, such as system updates win over user action updates.

There are 59 instances of calling updateCaseAndAssociations, so this refactoring effort might require a couple of weeks to remove these calls and verify our changes work as expected.

**Comment by codyseibert** [comment]
Talked with @zRogersFlexion about potential solutions.  We are not too sure about the proxy approach because it may go against our architecture design patterns and it seems a bit magic.  One potential solution would be to make very specific persistence methods that could be used to update a subset of properties on a case which would also make it more apparent what exactly is changing on the case when we invoke an interactor.  

A second approach mentioned was potentially putting all case updating logic behind a FIFO queue to allow each case modification to be made in order.  This approach is less appealing due to the extra infrastructure complexity it might create.  This approach may also slow down the performance of the system since every case change has to be pushed into a queue and processed one at a time.  So even if two api endpoints need to update unrelated fields on the case, they'd still be forced to wait.

**Comment by codyseibert** [comment]
spoke with @mmarcotte today and talked about some potential concerns and solutions.  He mentioned that @jimlerza suggests the "best" solution would probably involve adding some type of version to the entity and add conditional dynamodb writes.  This would cause our interactors to throw errors if we try to update entities that have already been updated since our copy.  We'd need to add logic to handle some type of conflict resolution (merge our changes into the upstream changes) and re-validate the entity before we save try to persist it.  We also talked about potentially having any PDF or email related operations occur only if all the dynamodb writes were successful.  This would prevent important documents from being modified or service emails from sending unless we know have confirmed data integrity.  Obviously adding all of this will involve a lot more refactoring and testing, so we'd need to discuss where we want this in the good, better, best options.  We also need to determine what happens if a long running process (practitioner updates their email) runs into one of these race conditions and is unable to merge during the conflict resolution steps.  We don't want the other cases in the process to prevent processing, but we'd need to keep track of which cases potentially failed, why, and setup a system to reprocess the cases.


We were in agreement that the updateCaseAndAssociations is the root cause to a lot of these issues and finding a solution that refactors away from that method would put the system into a better place.  The question really boils down to how much additional race condition checks should we add into place to prevent ALL race conditions from occurring, how much work might be involved, and how much we'd be willing to rework our solution if we defer adding some type of version check now.

**Comment by codyseibert** [comment]
accidentally pressed the close button

**Comment by codyseibert** [comment]
## Potential Solutions

### Ok

Remove all calls to updateCaseAndAssociations and instead invoke the individual updateCase, updateDocketEntry methods directly.  The logic in updateCaseAndAssociations seems buggy and the method is currently doing too much.  This won't prevent all race conditions, but this will probably fix some of the main ones.

### Good

This first solution would involve removing all instances of updateCaseAndAssocations in instead try to use patch methods to update individual fields on the entities so that we reduce the chances of overwriting previous changes.  This approach would still potentially have issues with race conditions, but these would be a lot more rare since often different interactors do not modify the same properties on entities.  Another issue with this approach is that there is a chance of causing entities to get into invalid states if we are not fully validating the entity before updating a subset of fields.  I've tried looking through the code to see if one of these race conditions is actually viable but it's hard to know.  Because of this, some entities may get into invalid states and cause our application to break.  This would mean we may need to stop throwing errors when entities are invalid when doing queries and instead just log fatal errors that we can investigate at a later time.  Right now our entire application will break on invalid entities until someone fixes the entity; instead, we could just have invalid data but log when our system finds something as being invalid.

This solution would probably require the less amount of work, but I think it would solve a majority of our race conditions, mainly because updateCaseAndAssocations is the main issue from what I'm seeing.

### Better

Instead of doing patch updates, we could instead track entity versions and do conditional expressions on our updates.  This would cause the update to fail if the upstream entity has changed at all.  We may also need to use dynamodb transactions in some situations where we update multiple entities in the interactor and we should treat these updates as a set of changes to the case.  Dynamodb transactions will automatically fail if any of the rows in dynamo are also in the process of being updated.  Transactions cost twice as much as normal writes, so we'd need to maybe keep that in mind.  

The issue with NOT doing transactions is we could potentially have case entities update but the docket entity might not update in some race condition scenarios.  

Another issue with this approach is we'd need to design write collision logic to figure out what we do when updating a dynamodb record fails.  For example, let's say two users try to update a docket entry at the same time?  There are two main concerns for collisions:
1. collision causes by users using the UI around the same time (2 tabs submit at the same time)
2. collisions causes by users + system logic modifying similar cases (1 user updates a case while a trial session is getting calendared)

I spoke to UX, they suggest we just show a modal that says something along the lines of "the record you're trying to update has been modified by another user" then give a button they can click to refresh their page.  This would solve issue 1. described above, but in regards to issue 2., we can't just have our system updates fail because a race condition.  The simple solution would be just have system logic always win in regards to race conditions and they overwrite any upstream data.  A more elegant solution would be write custom diffing logic to compare the old entity against the new entity, merge them together to prevent overwriting fields, check if any fields overlap and log if we force overwrote a field, then save the entity again.

We still need to keep in mind what happens if a race condition occurs during an interactor which needs to generate PDFs and Emails.  Some of our interactors may need to be refactored to do all PDF and Email service at the end so we 100% know the dynamo record was successfully updated.

### Best

All entities are versioned.  We use transactions for all writes.  All writes have some type of update condition (user deletes a party while another user tries to update a party).  Every interactor is refactored to do the PDF generation and Email service as the last part of the interactor logic (after all dynamodb updates were successful).  Users get a modal when a collision occurs that will tell them to refresh their page.  On write collision we do a smart diff to know what fields we actually need to update, refetch the new case, merge it with our chances, revalidate the entity, try to write again by just re-running the entire interactor logic.  If the entity gets into an invalid state and we can't correctly merge changes, we just need to throw an error.  We'd need to be cautious of the 29 second timeout when we start doing retries on the API.

Interactors are refactored in this type of way:

```javascript
function interactor() {
  await createPdfs()
  await updateDynamoRecordAndRetryOnCollisions()
  await sendServiceEmails()
}
```

might be some edge cases, for example, we don't know how many pages will be in a PDF until we create the PDF.  So we do end up having to do more dynamodb updates AFTER creating PDFs.

**Comment by codyseibert** [comment]
### Race condition scenario for collision resolution on versioned entities.

participant userA
participant updateCaseContextInteractor
participant dynamodb
participant updateDocketEntryMetaInteractor
participant userB
userA->updateCaseContextInteractor: changes status and title
userB->updateDocketEntryMetaInteractor: makes request
dynamodb->updateCaseContextInteractor: fetch case v1
dynamodb->updateDocketEntryMetaInteractor: fetch case v1
note over updateCaseContextInteractor: sets status to ready for trial
updateCaseContextInteractor->dynamodb: saves case (expect v1)
note over updateDocketEntryMetaInteractor: add docket entry to pending report
updateDocketEntryMetaInteractor->dynamodb: save case (expect v1)
note over updateDocketEntryMetaInteractor: conditional update fails
dynamodb->updateDocketEntryMetaInteractor: fetch case v2
note over updateDocketEntryMetaInteractor: refetch the case (v2)
note over updateDocketEntryMetaInteractor: conflict resolution\nmerge new fields into v2
note over updateDocketEntryMetaInteractor: revalidate case
note over updateDocketEntryMetaInteractor: if validation fails, how do we handle?
updateDocketEntryMetaInteractor-->>userB: if validation fails, \nkill request show modal and retry
updateDocketEntryMetaInteractor->dynamodb: save case (expect v2)
note over updateDocketEntryMetaInteractor: repeat conflict\nresolution as necessary

![Screen Shot 2023-01-23 at 10.57.04 AM.png](https://images.zenhubusercontent.com/5bbf6faf4368060246bbe5ce/c2c6dc44-9dcf-4b62-8aac-19bfd1860855)

**Comment by codyseibert** [comment]
here is a list of all interactors using the updateCaseAndAssocations with their line numbers:

```

shared/src/business/useCaseHelper/caseAssociation/associateIrsPractitionerToCase.js:
  51  
  52:     await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  53        applicationContext,

shared/src/business/useCaseHelper/caseAssociation/associatePrivatePractitionerToCase.js:
  70  
  71:     await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  72        applicationContext,

shared/src/business/useCaseHelper/docketEntry/fileAndServeDocumentOnOneCase.js:
  88      .getUseCaseHelpers()
  89:     .updateCaseAndAssociations({
  90        applicationContext,

shared/src/business/useCaseHelper/stampDisposition/addDraftStampOrderDocketEntryInteractor.js:
  107  
  108:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  109      applicationContext,

shared/src/business/useCases/addPetitionerToCaseInteractor.ts:
  57      .getUseCaseHelpers()
  58:     .updateCaseAndAssociations({
  59        applicationContext,

shared/src/business/useCases/archiveDraftDocumentInteractor.ts:
  55      .getUseCaseHelpers()
  56:     .updateCaseAndAssociations({
  57        applicationContext,

shared/src/business/useCases/blockCaseFromTrialInteractor.ts:
  45      .getUseCaseHelpers()
  46:     .updateCaseAndAssociations({
  47        applicationContext,

shared/src/business/useCases/checkForReadyForTrialCasesInteractor.ts:
  19      const caseEntity = entity.validate();
  20:     await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  21        applicationContext,

shared/src/business/useCases/prioritizeCaseInteractor.ts:
  56      .getUseCaseHelpers()
  57:     .updateCaseAndAssociations({
  58        applicationContext,

shared/src/business/useCases/removeCasePendingItemInteractor.ts:
  45  
  46:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  47      applicationContext,

shared/src/business/useCases/removePdfFromDocketEntryInteractor.ts:
  50        .getUseCaseHelpers()
  51:       .updateCaseAndAssociations({
  52          applicationContext,

shared/src/business/useCases/removePetitionerAndUpdateCaptionInteractor.ts:
  73      .getUseCaseHelpers()
  74:     .updateCaseAndAssociations({
  75        applicationContext,

shared/src/business/useCases/removeSignatureFromDocumentInteractor.ts:
  43  
  44:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  45      applicationContext,

shared/src/business/useCases/saveCaseDetailInternalEditInteractor.ts:
  160      .getUseCaseHelpers()
  161:     .updateCaseAndAssociations({
  162        applicationContext,

shared/src/business/useCases/saveSignedDocumentInteractor.ts:
  164  
  165:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  166      applicationContext,

shared/src/business/useCases/sealCaseContactAddressInteractor.ts:
  53      .getUseCaseHelpers()
  54:     .updateCaseAndAssociations({
  55        applicationContext,

shared/src/business/useCases/sealCaseInteractor.ts:
  35      .getUseCaseHelpers()
  36:     .updateCaseAndAssociations({
  37        applicationContext,

shared/src/business/useCases/unblockCaseFromTrialInteractor.ts:
  48      .getUseCaseHelpers()
  49:     .updateCaseAndAssociations({
  50        applicationContext,

shared/src/business/useCases/unprioritizeCaseInteractor.ts:
  59      .getUseCaseHelpers()
  60:     .updateCaseAndAssociations({
  61        applicationContext,

shared/src/business/useCases/unsealCaseInteractor.ts:
  35      .getUseCaseHelpers()
  36:     .updateCaseAndAssociations({
  37        applicationContext,

shared/src/business/useCases/updateCaseDetailsInteractor.ts:
  128      .getUseCaseHelpers()
  129:     .updateCaseAndAssociations({
  130        applicationContext,

shared/src/business/useCases/updateContactInteractor.ts:
  222    if (shouldUpdateCase) {
  223:     await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  224        applicationContext,

shared/src/business/useCases/updatePetitionerInformationInteractor.ts:
  292      .getUseCaseHelpers()
  293:     .updateCaseAndAssociations({
  294        applicationContext,

shared/src/business/useCases/updateQcCompleteForTrialInteractor.ts:
  46      .getUseCaseHelpers()
  47:     .updateCaseAndAssociations({
  48        applicationContext,

shared/src/business/useCases/caseAssociation/deleteCounselFromCaseInteractor.ts:
  64      .getUseCaseHelpers()
  65:     .updateCaseAndAssociations({
  66        applicationContext,

shared/src/business/useCases/caseAssociation/updateCounselOnCaseInteractor.ts:
  85      .getUseCaseHelpers()
  86:     .updateCaseAndAssociations({
  87        applicationContext,

shared/src/business/useCases/caseConsolidation/addConsolidatedCaseInteractor.ts:
  90      updateCasePromises.push(
  91:       applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  92          applicationContext,

shared/src/business/useCases/caseConsolidation/removeConsolidatedCasesInteractor.ts:
   66        updateCasePromises.push(
   67:         applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
   68            applicationContext,

   80      updateCasePromises.push(
   81:       applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
   82          applicationContext,

  105      updateCasePromises.push(
  106:       applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  107          applicationContext,

shared/src/business/useCases/caseDeadline/createCaseDeadlineInteractor.ts:
  53  
  54:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  55      applicationContext,

shared/src/business/useCases/caseDeadline/deleteCaseDeadlineInteractor.ts:
  50      .getUseCaseHelpers()
  51:     .updateCaseAndAssociations({
  52        applicationContext,

shared/src/business/useCases/caseNote/deleteCaseNoteInteractor.ts:
  36      .getUseCaseHelpers()
  37:     .updateCaseAndAssociations({
  38        applicationContext,

shared/src/business/useCases/caseNote/saveCaseNoteInteractor.ts:
  43      .getUseCaseHelpers()
  44:     .updateCaseAndAssociations({
  45        applicationContext,

shared/src/business/useCases/caseStatistics/addDeficiencyStatisticInteractor.ts:
  81      .getUseCaseHelpers()
  82:     .updateCaseAndAssociations({
  83        applicationContext,

shared/src/business/useCases/caseStatistics/deleteDeficiencyStatisticInteractor.ts:
  35      .getUseCaseHelpers()
  36:     .updateCaseAndAssociations({
  37        applicationContext,

shared/src/business/useCases/caseStatistics/updateDeficiencyStatisticInteractor.ts:
  85      .getUseCaseHelpers()
  86:     .updateCaseAndAssociations({
  87        applicationContext,

shared/src/business/useCases/caseStatistics/updateOtherStatisticsInteractor.ts:
  42      .getUseCaseHelpers()
  43:     .updateCaseAndAssociations({
  44        applicationContext,

shared/src/business/useCases/correspondence/archiveCorrespondenceDocumentInteractor.ts:
  51  
  52:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  53      applicationContext,

shared/src/business/useCases/courtIssuedOrder/fileCourtIssuedOrderInteractor.ts:
  120  
  121:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  122      applicationContext,

shared/src/business/useCases/courtIssuedOrder/updateCourtIssuedOrderInteractor.ts:
  126      .getUseCaseHelpers()
  127:     .updateCaseAndAssociations({
  128        applicationContext,

shared/src/business/useCases/docketEntry/addPaperFilingInteractor.ts:
  180  
  181:     await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  182        applicationContext,

shared/src/business/useCases/docketEntry/editPaperFilingInteractor.ts:
  257  
  258:     await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  259        applicationContext,

shared/src/business/useCases/docketEntry/fileCourtIssuedDocketEntryInteractor.ts:
  164        const saveItems = [
  165:         applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  166            applicationContext,

shared/src/business/useCases/docketEntry/updateCourtIssuedDocketEntryInteractor.ts:
   99      }),
  100:     applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  101        applicationContext,

shared/src/business/useCases/editDocketEntry/completeDocketEntryQCInteractor.ts:
  348  
  349:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  350      applicationContext,

shared/src/business/useCases/externalDocument/fileExternalDocumentForConsolidatedInteractor.ts:
  248            .getUseCaseHelpers()
  249:           .updateCaseAndAssociations({
  250              applicationContext,

shared/src/business/useCases/externalDocument/fileExternalDocumentInteractor.ts:
  174  
  175:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  176      applicationContext,

shared/src/business/useCases/serveCaseToIrs/serveCaseToIrsInteractor.ts:
  578  
  579:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  580      applicationContext,

shared/src/business/useCases/trialSessions/addCaseToTrialSessionInteractor.ts:
  88      .getUseCaseHelpers()
  89:     .updateCaseAndAssociations({
  90        applicationContext,

shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.ts:
  89  
  90:     await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  91        applicationContext,

shared/src/business/useCases/trialSessions/generateNoticesForCaseTrialSessionCalendarInteractor.ts:
  309  
  310:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  311      applicationContext,

shared/src/business/useCases/trialSessions/removeCaseFromTrialInteractor.ts:
  98      .getUseCaseHelpers()
  99:     .updateCaseAndAssociations({
  100        applicationContext,

shared/src/business/useCases/trialSessions/setTrialSessionCalendarInteractor.ts:
   31  
   32:   return applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
   33      applicationContext,

  129        }),
  130:       applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  131          applicationContext,

  155        }),
  156:       applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  157          applicationContext,

shared/src/business/useCases/trialSessions/updateTrialSessionInteractor.ts:
  93  
  94:     await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  95        applicationContext,

shared/src/business/useCases/users/generateChangeOfAddress.ts:
  120          .getUseCaseHelpers()
  121:         .updateCaseAndAssociations({
  122            applicationContext,

shared/src/business/useCases/users/verifyUserPendingEmailInteractor.ts:
  107      filteredCasesToUpdateInPersistence.map(caseToUpdate =>
  108:       applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  109          applicationContext,

  203      const validatedCaseToUpdate = validCasesToUpdate[idx];
  204:     await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  205        applicationContext,

shared/src/business/useCases/workitems/completeWorkItemInteractor.ts:
  84  
  85:   await applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
  86      applicationContext,

```

**Comment by codyseibert** [comment]
Another approach I've been trying is to setup lock entries in our dynamo table which are used to acquire a lock on a docker-number.  Basically any endpoint which attempts to modify something on a case will acquire the lock if one doesn't exist, or it will throw an exception and show a modal to the user that they need to refresh their page.  The lock has an expiration date which will cause it to be deleted if another interactors tries to acquire the lock after it's expired.  This helps prevent dead locks, but figuring out the best TTL might be hard.  This approach should work fine assuming most operations on a case are pretty fast.

**Comment by codyseibert** [comment]
Mike Jim and I talked today about potentially adding a similar locking approach as talked about in the comment above, but instead of having the failure to acquire a lock throw an error, we'd instead just console log an error so that we can see how prevalent this issue is in our system.  We don't really know how often users are trying to modify the same case at the same time, and implementing a solution to make this system 100% safe from concurrent modification will not be an easy task.

We also talked more about maybe focusing in on the updateCaseAndAssocations method to make it smarter so that it doesn't revert records in the database that the interactor didn't even modify.  For example, if someone is modifying docket entry 2, it should not consider touching docket entry 3 when it does it's diff logic.

**Comment by codyseibert** [comment]
I went down another prototype to just find a way to make updateCaseAndAssociations dumb.  Right now it's doing too much logic, and the main area of concern are the functions that run the `diff` check.

```
updateCaseDocketEntries,
updateCorrespondence,
updateHearings,
updateIrsPractitioners,
updatePrivatePractitioners
```

these functions compare the current full case against the full case in dynamo which causes it to overwrite changes to docket records unrelated to the docket record the interactor is attempting to update.  I would argue that this is the first area of concern we should refactor and try to make this function be passed in an array of ids it should only update instead of trying to diff and update everything.  This would not 100% prevent race conditions, but it's a small change which fixes the original bug described in this story and will probably prevent other similar bugs.

pr: https://github.com/ustaxcourt/ef-cms/pull/3086/files

**Comment by codyseibert** [comment]
for future reference, some work was done on the following branches

https://github.com/ustaxcourt/ef-cms/pull/3046
https://github.com/ustaxcourt/ef-cms/pull/3078
https://github.com/ustaxcourt/ef-cms/pull/3086
